### ISSUE TITLE
Payment object still present after applying 100% promotion coupon

### ISSUE THREAD

**Issue by Persata** [Sylius/Sylius#7952]
Hello all,

I've encountered an issue with promotions when reducing the order total to 0. The system is still creating a payment entry in the database, with an incorrect amount, and leaving the order as `awaiting_payment` even though there is no amount to pay.

This occurs when you have a promotion that is "Coupon based" which is then applied to the cart, as the `OrderPaymentProcessor` performs no operation in this case:

```php
    public function process(BaseOrderInterface $order)
    {
        /** @var OrderInterface $order */
        Assert::isInstanceOf($order, OrderInterface::class);

        if (OrderInterface::STATE_CANCELLED === $order->getState()) {
            return;
        }
        
        if (0 === $order->getTotal()) {
            return;
        }

        // ...
```

The current Behat tests for this scenario create a promotion that is _not_ coupon based,
so it applies as soon as the product is added to the cart - the `OrderPaymentProcessor` then does not create a payment in the first place (see the no-op above), allowing the tests to pass.

In the scenario **with** the coupon, the payment is created initially when the product is first added to the cart, with the value of the initial product. After adding the coupon-based promotion, the initial payment is left untouched with the initial value, regardless of whether you add further products to the cart.

I have created a failing test here to demonstrate: https://github.com/Persata/Sylius/commit/13e079ac0107df6356cb5850d2980e6945791e27

Updating the `OrderPaymentProcessor` method to the following removes the payment from the database when one is present AND the order total is 0:

```php
public function process(BaseOrderInterface $order)
    {
        /** @var OrderInterface $order */
        Assert::isInstanceOf($order, OrderInterface::class);

        if (OrderInterface::STATE_CANCELLED === $order->getState()) {
            return;
        }

        $lastPayment = $order->getLastPayment($this->targetState);

        if (null !== $lastPayment) {
            if (0 === $order->getTotal()) {
                $order->removePayment($lastPayment);
                return;
            } else {
                $lastPayment->setCurrencyCode($order->getCurrencyCode());
                $lastPayment->setAmount($order->getTotal());
                return;
            }
        }

        if (0 === $order->getTotal()) {
            return;
        }

        try {
            $newPayment = $this->orderPaymentProvider->provideOrderPayment($order, $this->targetState);
            $order->addPayment($newPayment);
        } catch (NotProvidedOrderPaymentException $exception) {
            return;
        }
    }
```

However even after completing the order with this, the order payment state is stuck at `awaiting_payment` since there is no payment to mark as complete.

I have looked into adding an additional order `payment_state` of "Not required" for orders with no payments and a total of 0, and using the state machine to update the `payment_state` of the order to this new value:

![not_required_payment](https://cloud.githubusercontent.com/assets/1608834/25141693/387dadb0-245c-11e7-82fe-3afd4ffb701b.png)

However a lot of extra logic will need modified to handle other state machine transitions when this payment state applies (e.g. cancellation, fulfilment etc)

Should the order's payment state in this case just simply be marked as "Paid"? I can create pull requests for either solution if they seem the ideal routes, but wanted to ask for input beforehand.

Sorry for the exceptionally long issue! Any thoughts or ideas?

Thanks!

**Comment by lchrusciel** [comment]
It is partially related to #7943. The similar problem exists for skipped shipping.

**Comment by pjedrzejewski** [comment]
Fixed in #8532, thanks for reporting Ross! :)
